src/Engine/ProtoCore/FFI/CLRObjectMarshaler.cs

```diff
@@ -1105,20 +1108,27 @@ private StackValue CreateDSObject(object obj, ProtoCore.Runtime.Context context,
         {
             //We are here, because we want to create DS object of user defined type.
             var runtimeCore = dsi.runtime.RuntimeCore;
-            var classTable = runtimeCore.DSExecutable.classTable;
             Type objType = GetPublicType(obj.GetType());
-            int type = classTable.IndexOf(GetTypeName(objType));
-            //Recursively get the base class type if available.
-            while (type == -1 && objType != null)
+
+            if (mCachedObjType != objType)
             {
-                objType = objType.BaseType;
-                if (null != objType)
-                    type = classTable.IndexOf(GetTypeName(objType));
+                mCachedObjType = objType;
+
+                var classTable = runtimeCore.DSExecutable.classTable;
+                mCachedType = classTable.IndexOf(GetTypeName(objType));
+                
+                //Recursively get the base class type if available.
+                while (mCachedType == -1 && objType != null)
+                {
+                    objType = objType.BaseType;
+                    if (null != objType)
+                        mCachedType = classTable.IndexOf(GetTypeName(objType));
```

Comment:
so... why are we doing this, why do we want to get the real base type?

Reply:
I don't know specifically.  It is for un-marshaling the object later though.  This data is put in the the `StackValue` object pointer